### PR TITLE
MGMT-12759: Reinstate assisted NMState Go binding

### PR DESCRIPTION
**Pull request by nmagnezi**
Basing on RHEL8.7 RPMs, nmstate 1.3.3

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by nmagnezi** [comment]
/hold
need CI results, but still WIP.

File: pkg/staticnetworkconfig/generator_test.go [hunk 1]
```diff
@@ -20,7 +20,9 @@ func TestStaticNetworkConfig(t *testing.T) {
 var _ = Describe("StaticNetworkConfig", func() {
 	var (
 		staticNetworkGenerator = StaticNetworkConfigGenerator{log: logrus.New()}
+		mockNMState            = MockNmstate{}
```
Comment (carbonin) [inline pkg/staticnetworkconfig/generator_test.go:23]: Seems strange that we're not actually using this anywhere in these tests.

Given the instability we've seen in nmstate wouldn't it be good to not mock this and actually validate that it's working correctly in these tests?

Maybe we have something like that in subsystem already?

File: pkg/staticnetworkconfig/generator_test.go [hunk 2]
```diff
@@ -20,7 +20,9 @@ func TestStaticNetworkConfig(t *testing.T) {
...
 	)
+	staticNetworkGenerator.nmstate = &mockNMState
```
[the remark this answers is not in the document]
Comment (nmagnezi) [reply pkg/staticnetworkconfig/generator_test.go]: Mock removed.

File: pkg/staticnetworkconfig/generator.go [hunk 3]
```diff
@@ -5,24 +5,24 @@ import (
...
-	"golang.org/x/sync/semaphore"
 	"gopkg.in/ini.v1"
 	"gopkg.in/yaml.v2"
 	"k8s.io/apimachinery/pkg/util/json"
+	yamlconvertor "sigs.k8s.io/yaml"
 )
 
-type Config struct {
-	MaxConcurrentGenerations int64 `envconfig:"MAX_CONCURRENT_NMSTATECTL_GENERATIONS" default:"30"`
+//go:generate mockgen -source=generator.go -package=staticnetworkconfig -destination=mock_generator.go
```
Comment (carbonin) [inline pkg/staticnetworkconfig/generator.go:23]: You don't need this either now, right?

**Review by carbonin — COMMENTED** [review]
Since we don't need to shell out to run `nmstatectl` now, we should be able to write unit-tests that run `generateConfiguration`.

Is that something we want to do? I know it would mean asserting we know the exact behavior of nmstate, but maybe a basic test where we just ensure it runs without error would be a good start?

**Comment by nmagnezi** [comment]
Changing this back to DRAFT while debugging `ci/prow/e2e-agent-compact `

File: pkg/staticnetworkconfig/generator.go (hunk 3 above)
Comment (nmagnezi) [reply pkg/staticnetworkconfig/generator.go]: removed.

**Comment by nmagnezi** [comment]
> Since we don't need to shell out to run `nmstatectl` now, we should be able to write unit-tests that run `generateConfiguration`.
> 
> Is that something we want to do? I know it would mean asserting we know the exact behavior of nmstate, but maybe a basic test where we just ensure it runs without error would be a good start?

we do test things in subsystem, however, I did add some basic unit tests to make sure it works as expected.

File: pkg/staticnetworkconfig/generator.go [hunk 4]
```diff
@@ -91,41 +85,22 @@ func (s *StaticNetworkConfigGenerator) generateHostStaticNetworkConfigData(ctx c
...
 	if err != nil {
-		s.log.WithError(err).Errorf("Failed to create temp file")
 		return "", err
 	}
-	defer func() {
-		f.Close()
-		os.Remove(f.Name())
-	}()
-	_, err = f.WriteString(hostYAML)
+	logrus.Infof("hostJSON: %s", hostJSON)
```
Comment (carbonin) [inline pkg/staticnetworkconfig/generator.go:96]: Remove debug logs.

File: pkg/staticnetworkconfig/generator.go [hunk 5]
```diff
@@ -91,41 +85,22 @@ func (s *StaticNetworkConfigGenerator) generateHostStaticNetworkConfigData(ctx c
...
-	}
-	stdout, stderr, retCode := executer.ExecuteWithContext(ctx, "nmstatectl", "gc", f.Name())
-	if retCode != 0 {
-		s.log.Errorf("<nmstatectl gc> failed, errorCode %d, stderr %s, input yaml <%s>", retCode, stderr, hostYAML)
-		errMsg := strings.Split(stderr, "Error:")
-		return "", fmt.Errorf("failed to execute 'nmstatectl gc', error: %s", strings.TrimSpace(errMsg[len(errMsg)-1]))
+		s.log.WithError(err).Errorf("nmstate GenerateConfiguration failed, input yaml <%s>", hostYAML)
+		return "", fmt.Errorf("nmstate GenerateConfiguration failed, error: %s", err.Error())
 	}
+	logrus.Infof("after")
```
Comment (carbonin) [inline pkg/staticnetworkconfig/generator.go:102]: Same this should be removed.

File: pkg/staticnetworkconfig/generator_test.go [hunk 6]
```diff
@@ -17,6 +18,39 @@ func TestStaticNetworkConfig(t *testing.T) {
 	RunSpecs(t, "StaticNetworkConfig Suite")
 }
 
+var _ = Describe("Generate Config", func() {
```
Comment (carbonin) [inline pkg/staticnetworkconfig/generator_test.go:21]: Nit, but can you make this the name of the method you're testing? i.e. `Describe("generateConfiguration"...`

Comment (nmagnezi) [reply pkg/staticnetworkconfig/generator_test.go]: done

File: pkg/staticnetworkconfig/generator.go (hunk 5 above)
Comment (nmagnezi) [reply pkg/staticnetworkconfig/generator.go]: Yikes. Removed.

File: pkg/staticnetworkconfig/generator.go (hunk 4 above)
Comment (nmagnezi) [reply pkg/staticnetworkconfig/generator.go]: Yikes. Removed.

**Review by carbonin — APPROVED** [review]

**Pull request merged**